### PR TITLE
Support for more streams types (icecast, shoutcast, custom)

### PR DESCRIPTION
**Pull request by confact**
I removed status code and works with checking metaint exists instead.
Also had issues with the header parser and fixed it

This works with most of my streams now.

This needs of course some ruby-ifying.

### CODE REVIEW

**Comment by confact** [comment]
I get some checkups that is like infinite looking for a metadata. I think i need a timeout somehow in the metadata lookup in the read_metadata function

File: lib/shoutout/headers.rb [hunk 1]
```diff
@@ -2,9 +2,9 @@ module Shoutout
   class Headers < Hash
     def self.parse(raw_headers)
       headers = {}
-      raw_headers.split($/).each do |line|
+      raw_headers.split("\r\n").each do |line|
         key, value = line.chomp.split(":", 2)
-        headers[key.strip] = value.strip
```
Comment (DouweM) [inline lib/shoutout/headers.rb:7]: Why was this change necessary?

File: lib/shoutout/stream.rb [hunk 2]
```diff
@@ -34,32 +34,12 @@ def connect
 
       uri = URI.parse(@url)
       @socket = TCPSocket.new(uri.host, uri.port)
-      @socket.puts "GET #{uri.path} HTTP/1.0"
-      @socket.puts "Icy-MetaData: 1"
-      @socket.puts
-
-      # Read status line
-      status_line = @socket.gets
-      status_code = status_line.match(/\AHTTP\/([0-9]\.[0-9]) ([0-9]{3})/)[2].to_i
```
Comment (DouweM) [inline lib/shoutout/stream.rb:43]: Why was status code parsing removed entirely? This prevents shoutout from working with redirects and errors like "404 Not Found".

Comment (confact) [inline lib/shoutout/stream.rb:43]: This didn't work at all with most of my streams. I tested the status_code fix against 3k stations and it worked for 3-5 stations.(mostly because they where normal websites sending 404). This could of course be fixed somehow. But after removing this and fixed the header parser most of my stations started to work.

File: lib/shoutout/stream.rb [hunk 3]
```diff
@@ -71,6 +51,12 @@ def connect
       true
     end
 
+    def send_header_request(address)
+        return "GET / HTTP/1.1\r\nHost: #{address}\r\nConnection: close\r\n" +
+               "icy-metadata: 1\r\ntransferMode.dlna.org: Streaming\n\r\nHEAD / HTTP/1.1\r\n" +
+               "Host: #{address}\r\n" + "User-Agent: DirbleScrobbler\n\r\n";
+    end
```
Comment (DouweM) [inline lib/shoutout/stream.rb]: Why do we first send a GET request and then a HEAD request?

What is the `transferMode.dlna.org` header and why is it necessary?

Why are the `Host`, `Connection` and `User-Agent` headers necessary?

Comment (confact) [inline lib/shoutout/stream.rb]: This was a test for myself i accidentally commited. It should not be necessary. I can remove this and try it again.

File: lib/shoutout/stream.rb (hunk 2 above)
Comment (DouweM) [inline lib/shoutout/stream.rb:43]: Could you give me an example (or multiple) of streams that don't work properly with status code parsing (which is just regular HTTP, which the streams should handle to just fine, but oh well).

File: lib/shoutout/stream.rb (hunk 3 above)
Comment (DouweM) [inline lib/shoutout/stream.rb]: Please do.

File: lib/shoutout/headers.rb (hunk 1 above)
Comment (confact) [inline lib/shoutout/headers.rb:7]: i got wrong headers here all the time because of the split, it couldn't split the lines on $/ but \r\n worked. I removed strip because it should not have any spaces or \r\n now anymore after that change.

File: lib/shoutout/stream.rb (hunk 2 above)
Comment (confact) [inline lib/shoutout/stream.rb:43]: I run a kinda working shoutout now so I don't want to ruin it.

But i guess all or some of these didn't work before the status code and header bug:
http://stream.liffeysoundfm.ie:8000/
http://nl1.streamhosting.ch:80
http://s8.voscast.com:8474/;
http://91.121.82.169:8072
http://206.190.150.90:8415
http://200.43.192.20:8010
http://189.1.162.125:8450
http://91.121.147.57:8100
http://87.229.45.44:8400

These now works but hangs when getting metadata.

File: lib/shoutout/headers.rb (hunk 1 above)
Comment (DouweM) [inline lib/shoutout/headers.rb:7]: If you have `Header: value`, `key` will now have `"Header"` and value `" value"` note the space in front!

File: lib/shoutout/stream.rb [hunk 4]
```diff
@@ -35,31 +35,13 @@ def connect
       uri = URI.parse(@url)
       @socket = TCPSocket.new(uri.host, uri.port)
       @socket.puts "GET #{uri.path} HTTP/1.0"
-      @socket.puts "Icy-MetaData: 1"
+      @socket.puts "icy-metadata: 1"
```
Comment (DouweM) [inline lib/shoutout/stream.rb:38]: Header parsing by servers is supposed to be case insensitive—were your servers choking on the camel cased version?

Comment (confact) [inline lib/shoutout/stream.rb:38]: Oh, no. I am just used to type icy-metadata in lower case.

File: lib/shoutout/headers.rb (hunk 1 above)
Comment (confact) [inline lib/shoutout/headers.rb:7]: Okay, then I think i found another bug. I will see if I am right and commit it to this branch when tested.

Comment (confact) [inline lib/shoutout/headers.rb:7]: fixed, just checking if key and value is not nil.

Comment (DouweM) [inline lib/shoutout/headers.rb:7]: In what situation are you getting nil for key or value? That would indicate a malformed header line...

Comment (confact) [inline lib/shoutout/headers.rb:7]: This could be status code causing malformed header when i removed it in connect. It will probably follow to the header then.

Comment (DouweM) [inline lib/shoutout/headers.rb:7]: Yeah, that makes sense. I won't need this change then, as I'll attempt to fix rather than remove the status code check.

**Comment by DouweM** [comment]
Thanks for your help!

Looks like the only real issue here is that some servers don't play well with the HTTP status code check. I'll investigate that issue when I find time and will attempt to incorporate a proper fix.

I'll let you know when that's done so you can get back on the canonical release of shoutout rather than your own fork.

**Comment by confact** [comment]
The code right now get stuck all the time in read_metadata. Could this be because it is looking and looking but can't find metadata? Shall we try add a Timeout or add a max retries before just say we couldn't find a song?

That could break your listen function though. What to do?

**Comment by DouweM** [comment]
How are you using shoutout right now? `#read_metadata` is run in a separate thread and is supposed to be an infinite loop (until the socket is closed, of course).

(Adding a timeout for when the socket isn't responding properly isn't a bad idea, I just want to make sure we're talking about the same thing)

**Comment by confact** [comment]
Now I am just scrobbling thousands of streams for metadata in a worker/cronjob and want to disconnect it asap. I want it to go as fast as it can. But when streams says they will have metadata but won't, it get stuck there and just listen forever.

code example:

```
status = 0
stream = Shoutout::Stream.new(url)
begin
  okay = stream.connect
  if okay == true
    artist = stream.metadata.artist
    song = stream.metadata.song
    stream.disconnect
    status = 1
  else
    song = ""
    artist = ""
  end
rescue Errno::ECONNREFUSED
  @stream = Stream.find_by_stream(url)
  @stream.update(status:0)
  return false
rescue Errno::ETIMEDOUT
  @stream = Stream.find_by_stream(url)
  @stream.update(status:0)
  return false
rescue SocketError
  @stream = Stream.find_by_stream(url)
  @stream.update(status:0)
  return false
rescue Errno::ECONNRESET
  return false
rescue URI::InvalidURIError
  return false
end
```

I will use bitrate and content_type in a different worker.

**Comment by confact** [comment]
I run the worker with some print to know where it get stuck.read headers,done is just checking so it isn't parse_headers. in read metadata is inside read_metadata - this is how it looks when i run sidekiq: 

```
2014-12-09T11:14:04.535Z 12393 TID-1aqsvo GetStationSongWorker JID-6078ce311ffa017bd183b211 INFO: start
2014-12-09T11:14:04.535Z 12393 TID-1aqqt8 GetStationSongWorker JID-effecbaad1a4e5f98b11d79b INFO: start
read headers, done.in read metadataread headers, done.in read metadataread headers, done.in read  metadataread headers, done.read headers, done.in read metadatain read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadataread headers, done.in read metadata`
```

this show it have been listen for over one hour now, all 25 workers are stuck.

**Comment by DouweM** [comment]
Hmm, that's interesting. By "But when streams says they will have metadata but won't, it get stuck there and just listen forever.", do you mean connecting goes through and the server returns an `icy-metaint` header, but the metadata length is always 0 (see [lib/shoutout/stream.rb#L139](https://github.com/DouweM/shoutout/blob/master/lib/shoutout/stream.rb#L139)), i.e. the server isn't sending any metadata? Because you're right, in that case `#metadata` would stall, waiting until the server finally returns some metadata.

When `#listen`ing, we would only want to fail when the server stops sending anything at all, because we would be content waiting for the "next change", but you're right: `#metadata` should timeout if the server isn't sending any metadata in a timely manner.

I'll look into that as well, but you could easily work it into your code by replicating and slightly changing `#metadata` behaviour:

``` ruby
received = false
stream.metadata_change do |new_metadata|
  received = true
end

tries = 0
until received
  # We want to give up after 1.5 seconds, i.e. 100 tries
  break if tries == 100

  sleep 0.015 

  tries += 1
end

if received
  # metadata has been received so you can safely use `stream.metadata`
else
  # nope :(
end
```

**Comment by DouweM** [comment]
Is there any good reason for these streams to not send any metadata? Are they broken or is shoutout?

**Comment by confact** [comment]
Oh thanks! This is a normal way. every stream is sending a metaint when icy-metadata is set to 1 as default, but that doesn't mean they need to give any metadata at all, ever. So yea, right now some streams will stall because of this.

thanks for the code. I will use that. Now i have changed the while loop in stream.rb to go around this. this will probably work better than that, it still stall sometimes now.

**Comment by DouweM** [comment]
All right, thanks. This gem was mostly built for my own convenience, and only tested with the couple of streams I listen to or could easily find for Dutch radio stations. As you'll have noticed, it isn't exactly "production-ready" yet, so I'm glad you're putting it through some more rigorous testing to make it more robust!

**Comment by confact** [comment]
I know, I like to live on the edge. Your is much better than my own try to make a streamscrobbler gem. You're welcome. You have made amazing stuff with the gem and thanks for the inputs.

That received code worked as a charm. it goes faster than ever before. Now it is just some crashes due to my own code outside shoutout in my worker. So thanks once again!

**Comment by DouweM** [comment]
If you have any other shoutout or Ruby related questions, I'd be happy to help!

**Comment by confact** [comment]
I have fixed status code and other issues in my shoutout fork in branch fixing_hangs. Want me to do a pull_request?

**Comment by DouweM** [comment]
Please do.
